New syntax component: filter

annda opened the pull request:
This introduces a new component "filter" which accepts the same syntax as "table". It displays a form with available column values as radio buttons. It is built on top of the existing handling of dynamic search filters via `flt` GET parameters.

splitbrain commented:
There are some visual issues...

* It seems the header syntax is currently ignored
* HTML escaping seems to be wrong (double escaped)
* Wide filter names are missing the drop down element
* the alignment on multiline values isn't ideal - it would be nicer to have the second row indented, probably an easy fix using display:flex
* I would completely remove the fieldset and intro text
* It would be nice if filters would use the full available width (with some min-width and possibly a max-width too), again this might be achievable with flex and grow/shrink

![screenshot-localhost-2023 07 03-16_16_21](https://github.com/cosmocode/dokuwiki-plugin-struct/assets/86426/ec7518a3-ebe8-4bac-bd87-114a69b46dc7)

splitbrain commented:
The filters are currently radio boxes. This means I can only ever filter for one thing:

![screenshot-localhost-2023 07 03-16_29_43](https://github.com/cosmocode/dokuwiki-plugin-struct/assets/86426/6ff3a0fc-f527-4561-91bc-ad60002e4be4)

There is no way to filter for items matching "Schlüsselaktivitäten" OR "Kundenbeziehungen". I wonder if you implemented it this way because our dynfilters do not allow for multiple filters on the same field or if it was a deliberate choice on your part?

splitbrain commented on meta/Filter.php:
```diff
@@ -0,0 +1,130 @@
+            $i = 0;
+            foreach ($values as $value) {
+                $form->addTagOpen('li');
+                $form->addRadioButton(SearchConfigParameters::$PARAM_FILTER . "[$qualifiedColName*~]")
```
This uses a contains comparison, but since we have the exact values in a dropdown, an exact comparison should be better?

annda replied:
Theoretically you are right. But for some reason input is not processed correctly when the name contains `=` and automatic pre-filling does not work. So instead of handling the `checked` status on my own I decided to keep the `like' operand.

splitbrain commented on meta/Filter.php:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @return array
+     */
+    protected function getAllColumnValues(): array
```
Would it be useful to have a test for this method?

splitbrain commented on meta/Filter.php:
```diff
@@ -0,0 +1,130 @@
+        // column dropdowns
+        foreach ($colValues as $colName => $colData) {
+            $qualifiedColName = $colName[0] !== '%' ? "$schema.$colName" : $colName;
+            $values = array_unique($colData['values']);
```
I think getAllColumnValues() should return unique values already.

annda replied:
Done.

splitbrain commented on meta/Filter.php:
```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        return $colValues;
```
Besides ensuring uniqueness here (see above) it probably also makes sense to sort the values? Alphabetically? Or maybe be frequency?

annda replied:
Done.

splitbrain commented on script/StructFilter.js:
```diff
@@ -0,0 +1,7 @@
+ * remove toggling inherited by reusing core search form
+ */
+jQuery('fieldset.struct-filter-form.search-form button.toggleAssistant').remove();
+jQuery('fieldset.struct-filter-form.search-form .advancedOptions').show();
```
These come from reusing the  default search styles? Can we avoid adding and removing stuff here?

splitbrain commented on meta/Filter.php:
```diff
@@ -0,0 +1,130 @@
+/**
+ * Struct filter class
+ */
+class Filter
```
when rebasing on #644 this will need to inherit from `Aggregation`

splitbrain commented on syntax/filter.php:
```diff
@@ -0,0 +1,115 @@
+use dokuwiki\plugin\struct\meta\SearchConfig;
+use dokuwiki\plugin\struct\meta\StructException;
+
+class syntax_plugin_struct_filter extends DokuWiki_Syntax_Plugin
```
when rebasing on #644 this probably should inherit from `syntax_plugin_struct_table`  and then `$this->illegalOptions` should be configured accordingly.

splitbrain reviewed (commented):
This looks fine already, but could be improved. Please also rebase on #644 so we can merge both PRs together.

splitbrain commented:
Hmm. I'm confused. Didn't you rebase your branch on top of nestedlists? I changed the base of the PR but still get a bunch of commits that have nothing to do with your changes?

annda commented:
> HTML escaping seems to be wrong (double escaped)

This is a bug in the Form class, I will submit a PR later:

`$this->addLabelHTML(hsc($label), $for, $pos);` which escapes the content again: `$element = new LabelElement(hsc($content));`

annda commented:
> Wide filter names are missing the drop down element

Fixed.

> the alignment on multiline values isn't ideal - it would be nicer to have the second row indented, probably an easy fix using display:flex

Fixed.

annda commented:
> I would completely remove the fieldset and intro text

I removed the intro text. But I need the fieldset for the search styles and Javascript to work.
